core/src/main/java/kafka/server/share/SharePartitionManager.java

```diff
@@ -317,6 +364,197 @@ public void close() throws Exception {
+                                0,
+                                partitionMaxBytes,
+                                Optional.empty()
+                            )
+                        );
+                    } else {
+                        sharePartition.releaseFetchLock();
+                        log.info("Record lock partition limit exceeded for SharePartition with key {}, " +
+                            "cannot acquire more records", sharePartitionKey);
+                    }
+                }
+            });
+
+            if (topicPartitionData.isEmpty()) {
+                // No locks for share partitions could be acquired, so we complete the request and
+                // will re-fetch for the client in next poll.
+                shareFetchPartitionData.future.complete(Collections.emptyMap());
+                // Though if no partitions can be locked then there must be some other request which
+                // is in-flight and should release the lock. But it's safe to release the lock as
+                // the lock on share partition already exists which facilitates correct behaviour
+                // with multiple requests from queue being processed.
+                releaseProcessFetchQueueLock();
+                return;
+            }
+
+            log.trace("Fetchable share partitions data: {} with groupId: {} fetch params: {}",
+                topicPartitionData, shareFetchPartitionData.groupId, shareFetchPartitionData.fetchParams);
+
+            replicaManager.fetchMessages(
+                shareFetchPartitionData.fetchParams,
+                CollectionConverters.asScala(
+                    topicPartitionData.entrySet().stream().map(entry ->
+                        new Tuple2<>(entry.getKey(), entry.getValue())).collect(Collectors.toList())
+                ),
+                QuotaFactory.UnboundedQuota$.MODULE$,
+                responsePartitionData -> {
+                    log.trace("Data successfully retrieved by replica manager: {}", responsePartitionData);
+                    List<Tuple2<TopicIdPartition, FetchPartitionData>> responseData = CollectionConverters.asJava(
+                        responsePartitionData);
+                    processFetchResponse(shareFetchPartitionData, responseData).whenComplete(
+                        (result, throwable) -> {
+                            if (throwable != null) {
+                                log.error("Error processing fetch response for share partitions", throwable);
+                                shareFetchPartitionData.future.completeExceptionally(throwable);
+                            } else {
+                                shareFetchPartitionData.future.complete(result);
+                            }
+                            // Releasing the lock to move ahead with the next request in queue.
+                            releaseFetchQueueAndPartitionsLock(shareFetchPartitionData.groupId, topicPartitionData.keySet());
```

Comment:
Hmm. Doesn't this mean only a single fetch at a time is possible for the entire SharePartitionManager? The fetch queue lock is only released once the replica manager has returned the records. This strikes me as a bad thing.

Reply:
Moved it out and verified the fetch, done.

Reply:
I think there's still more work to do in this area, but of course this code is still under active development so this is OK for now.

Reply:
Is this issue resolved? `releaseFetchQueueAndPartitionsLock` is still called when the callback completes, not when `replicaManager.fetchMessages` returns.

Reply:
This has been done to prevent completing requests aggressively when majority requests are for same partition as partition should still be locked till replica manager fetch completes. However as this is an initial working implementation and we are in active develoment of re-defining fetches hence this area of code will change majorly in subsequent weeks.